src/firefly/test/edu/caltech/ipac/astro/IpacTableReaderTest.java

```diff
@@ -0,0 +1,801 @@
+ *
+ * The test cases below test if the reader can parse all parts in a table correctly and also if the reader can throw exceptions correctly
+ *      when a table violate the requirements.
+ *
+ */
+public class IpacTableReaderTest {
+
+    //Input parameters:
+    private static final String catName = "catName";
+    String[] onlyColumns = null;
+    private boolean useFloatsForDoubles = false;
+    private boolean isHeadersOnlyAllow = false;
+    private boolean noAttributes = false;
+    private int estFileSize = 0;
+
+    //Expected parameters in the data group:
+    private String[] attributeKeys = null;
+    private String[] attributeValues = null;
+    private String[] attributeComments = null;
+    private String[] colTitles = null;
+    private String[] colKeyNames = null;
+    private String[] colDataTypeDesc = null;
+    private String[] colDataTypes = null;
+    private String[] colDataUnits = null;
+    private String[] colNullStrings = null;
+    private String[][] dataValues = null;
+
+    //The IPAC table reader store the table content in the dataGroup:
+    private DataGroup dataGroup;
+
+    @Before
+    /**
+     *
+     */
+    public void setUp() throws IpacTableException, ClassNotFoundException, java.io.FileNotFoundException {
+        //Read different tables in each test case below.
+    }
+
+    @After
+    /**
+     * Release the memories
+     */
+    public void tearDown() {
+        dataGroup = null;
+    }
+
+
+    @Test
+    /**
+     * This test calls the methods IpacTableReader.readIpacTable to read all the columns from a valid input IPAC table, IPACTableSample1.tbl,
+     * and verify if the produced dataGroup contains the correct information from the table.
+     *
+     * The input parameters to call IpacTableReader.readIpacTable:
+     * File
+     * onlyColumns
+     * useFloatsForDoubles
+     * catName = "catName"
+     * isHeadersonlyAllow
+     */
+    public void testValidIpacTable1() throws IpacTableException, ClassNotFoundException, java.lang.NullPointerException {
+
+        //Read all the columns, other parameters are "false".
+
+        /* IPACTableSample1.tbl:
+
+        \catalog1 = 'Sample Catalog1'
+        \catalog2 = 'Sample Catalog2'
+        \ Comment1
+        \ Comment2
+        |   ra      |    dec    |   n_obs  |    V     |   SpType   |
+        |   double  |    double |   int    |   real   |   char     |
+        |   deg     |    deg    |          |   mag    |            |
+        |   null    |    null   |   null   |   null   |   null     |
+          165.466279  -34.704730      5       11.27       K6Ve
+          123.4       5.67            9       8.9         K6Ve-1
+
+        */
+
+        //Inputs:
+        String tableName = "IPACTableSample1.tbl";
+        File inFile = new File(FileLoader.getDataPath(IpacTableReaderTest.class) + tableName);
+        //onlyColumns = null;
+
+        //Set the expected values:
+        setExpectedValues_Input1_full();
+
```

Comment:
This @Test gives the impression that it's simple and small based on the amount of code in it.  However, it is not the case.  setExpectedValues_Input1_full() has key information that require investigation to understand.  It also rely on the input, therefore the comment is repeated again in this function.  Now, you need to keep in sync the 2 comments and the input file.

checkResult() is a share function which is clever but is relying on setExpectedValues_Input1_full() to fully set up its expected values.  The reuse is good, but the dependencies may lead to bugs.